File: packages/jexlang-java/src/main/java/com/jexlang/java/types/JexNumber.java
package com.jexlang.java.types;

public class JexNumber implements JexValue {
    private final Number value;
    public JexNumber(Number value) { this.value = value; }

    @Override
    public String getType() {
        return "number";
    }

    @Override
    public String toString() {
        if (value instanceof Double || value instanceof Float) {
            return String.format("%.2f", value.doubleValue());
        } else {
            return value.toString();
        }
    }

    @Override
    public Object toObject() {
        return value; // Represents the number value in Java
    }

    public boolean isInteger() { return false; }
    public boolean isDouble() { return false; }
    public boolean isNumber() { return true; }
    public boolean isBoolean() { return false; }
    public boolean isString() { return false; }
    public boolean isNull() { return false; }
    public boolean isArray() { return false; }
    public boolean isObject() { return false; }
    public Number asNumber(String ctx) { return value; }
    public Integer asInteger(String ctx) { throw JexValue.typeError("integer", ctx, this); }
    public Double asDouble(String ctx) { throw JexValue.typeError("double", ctx, this); }
    public boolean asBoolean(String ctx) { throw JexValue.typeError("boolean", ctx, this); }
    public String asString(String ctx) { throw JexValue.typeError("string", ctx, this); }
    public java.util.List<JexValue> asArray(String ctx) { throw JexValue.typeError("array", ctx, this); }
    public java.util.Map<String, JexValue> asObject(String ctx) { throw JexValue.typeError("object", ctx, this); }
}

File: packages/jexlang-swift/jexlang-swift/eval/types/JexNumber.swift
//
//  JexNumber.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 02/11/25.
//

import Foundation

public class JexNumber: JexValue {
    
    private let value: NSNumber
    
    public init(value: NSNumber) {
        self.value = value
    }
    
    public func isInteger() -> Bool {
        return false
    }
    
    public func isDouble() -> Bool {
        return false
    }
    
    public func isNumber() -> Bool {
        return true
    }
    
    public func isBoolean() -> Bool {
        return false
    }
    
    public func isString() -> Bool {
        return false
    }
    
    public func isNil() -> Bool {
        return false
    }
    
    public func isArray() -> Bool {
        return false
    }
    
    public func isObject() -> Bool {
        return false
    }
    
    public func getType() -> String {
        return "number"
    }
    
    public var description: String {
        return "\(value)"
    }
    
    public func toObject() -> AnyObject? {
        return value
    }
    
    public func asNumber(context: String) throws -> NSNumber {
        return value
    }
    
    public func asDouble(context: String) throws -> Double {
        throw JexValueFactory.typeError(want: "double", ctx: context, actualValue: self)
    }
    
    public func asInteger(context: String) throws -> Int {
        throw JexValueFactory.typeError(want: "integer", ctx: context, actualValue: self)
    }
    
    public func asBoolean(context: String) throws -> Bool {
        throw JexValueFactory.typeError(want: "boolean", ctx: context, actualValue: self)
    }
    
    public func asString(context: String) throws -> String {
        return description
    }
    
    public func asArray(context: String) throws -> [JexValue] {
        throw JexValueFactory.typeError(want: "array", ctx: context, actualValue: self)
    }
    
    public func asObject(context: String) throws -> [String : JexValue] {
        throw JexValueFactory.typeError(want: "object", ctx: context, actualValue: self)
    }
}
